src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +76,120 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Forward(
-    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+    const arma::Mat<eT>& input,
+    arma::Mat<eT>& output)
 {
-  // Mean and variance over the entire training set will be used to compute
-  // the forward pass when deterministic is set to true.
-  if (deterministic)
+  Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+      by feature maps.");
+
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = input.n_rows / size;
+
+  // Set size of output equal to the size of input.
+  output.set_size(arma::size(input));
+
+  // We will calculate minibatch norm on each channel / feature map.
+  if (!deterministic)
   {
-    // Normalize the input and scale and shift the output.
-    output = input.each_col() - runningMean;
-    output.each_col() %= gamma / arma::sqrt(runningVariance / count + eps);
-    output.each_col() += beta;
+    // Input corresponds to output from convolution layer.
+    // Use a cube for simplicity.
```

Comment:
Hey why did we remove the linear layer case from here? IMO we should have it for speed optimization

Reply:
It was getting complicated to manage both of them, any change we needed to make then they needed to made in both of them. This should be easier to maintain.

Reply:
can you also explain the reason for the serialisation issue?

Reply:
I think the reason was when we try to load the values there was a division by zero in the expression, It was resolved by add eps to the expression. input.n_elem / size (Both initialized to 0).

Reply:
>  IMO we should have it for speed optimization

I have made the changes for this as well, Let me know if I should add it or we can keep it as it is.